src/qttools/kernels/cuda/banded.py

```diff
@@ -0,0 +1,1071 @@
+
+        num_denseC_blocks = tl.cdiv(M, BLK_N)
+
+        bandA_col_start = max(0, denseC_col - denseC_row + diag_dist_a - diag_dist_b)
+        bandB_row_start = max(0, -denseC_col + denseC_row - diag_dist_a + diag_dist_b)
+
+        num_blocks = min(band_a - bandA_col_start, band_b - bandB_row_start)
+
+        # ----------------------------------------------------------
+        # Create pointers for the first blocks of A and B.
+        # We will advance this pointer as we move in the K direction
+        # and accumulate
+        # `a_ptrs` is a block of [BLOCK_SIZE_M, BLOCK_SIZE_K] pointers
+        # `b_ptrs` is a block of [BLOCK_SIZE_K, BLOCK_SIZE_N] pointers
+        # See above `Pointer Arithmetic` section for details
+        offs_am = (denseC_row * BLK_M + tl.arange(0, BLK_M)) % M
+        offs_bn = (denseC_col * BLK_N + tl.arange(0, BLK_N)) % M
+        offs_k = tl.arange(0, BLK_K)
+        a_ptrs = a_ptr + (offs_am[:, None] * stride_am + offs_k[None, :] * stride_ak)
+        b_ptrs = b_ptr + (offs_k[:, None] * stride_bk + offs_bn[None, :] * stride_bn)
+
+        if denseC_col >= num_denseC_blocks:
+            return
+        if denseC_row >= num_denseC_blocks:
+            return
+        if denseC_col < 0:
+            return
+        if denseC_row < 0:
+            return
+
+        # # move the pointers to the start of the band
+        a_ptrs += bandA_col_start * BLK_N * stride_ak
+        b_ptrs += bandB_row_start * BLK_M * stride_bk
+
+        a_ptr += bandA_col_start * BLK_N * stride_ak
+        b_ptr += bandB_row_start * BLK_M * stride_bk
+
+        # -----------------------------------------------------------
+        # Iterate to compute a block of the C matrix.
+        # We accumulate into a `[BLOCK_SIZE_M, BLOCK_SIZE_N]` block
+        # of fp32 values for higher accuracy.
+        # `accumulator` will be converted back to fp16 after the loop.
+
+        accumulator = tl.zeros((BLK_M, BLK_N), dtype=tl.float64)
+
+        for k in range(0, tl.cdiv(num_blocks * BLK_M, BLK_K)):
+            a = tl.load(
+                a_ptrs
+            )  # , mask=offs_k[None, :] < band_a - k * BLK_K, other=0.0)
+            b = tl.load(
+                b_ptrs
+            )  # , mask=offs_k[:, None] < band_b - k * BLK_K, other=0.0)
+            # We accumulate along the K dimension.
+            for k_idx in range(BLK_K):
+                # Calculate pointers for the k_idx-th column of a and row of b
+                a_col_ptrs = (
+                    a_ptr + offs_am * stride_am + (k_idx + k * BLK_K) * stride_ak
+                )
+                b_row_ptrs = (
+                    b_ptr + (k_idx + k * BLK_K) * stride_bk + offs_bn * stride_bn
+                )
+
+                # Load the column and row
+                a_col = tl.load(a_col_ptrs)  # [BLOCK_SIZE_M]
+                b_row = tl.load(b_row_ptrs)  # [BLOCK_SIZE_N]
+
+                # Outer product accumulation
+                accumulator += tl.where(
```

Comment:
Aren't `offs_am` and `offs_bn` by definition (lines 905-906) less than M? Is it possible that this is killing performance?